codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,297 @@
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState state) {
+        state(state.stream()).updateStreamableBytes(state.streamableBytes(),
+                                                    state.hasFrame() && state.isWriteAllowed());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        while (connectionState.isActiveCountForTree != 0 && maxBytes > 0) {
+            // connectionState will never have a frame to send, so go right to children.
+            maxBytes -= distributeToChildren(maxBytes, writer, connectionState);
+        }
+
+        return connectionState.isActiveCountForTree != 0;
+    }
+
+    /**
+     * Sets the amount of bytes that will be allocated to each stream. Defaults to 1KiB.
+     * @param minAllocationChunk the amount of bytes that will be allocated to each stream. Must be > 0.
+     */
+    public void allocationQuantum(int allocationQuantum) {
+        if (allocationQuantum <= 0) {
+            throw new IllegalArgumentException("allocationQuantum must be > 0");
+        }
+        this.allocationQuantum = allocationQuantum;
+    }
+
+    private int distribute(int maxBytes, Writer writer, State pState) throws Http2Exception {
+        if (pState.isActive) {
+            int nsent = min(maxBytes, pState.streamableBytes);
+            pState.write(nsent, writer);
+            if (nsent == 0) {
+                // If a stream sends zero bytes, then we gave it a chance to write empty frames and it is now
+                // considered inactive until the next call to updateStreamableBytes. This allows descendant streams to
+                // be allocated bytes when the parent stream can't utilize them. This may be as a result of the
+                // stream's flow control window being 0.
+                pState.updateStreamableBytes(pState.streamableBytes, false);
+                return 0;
+            }
+            connectionState.pseudoTime += nsent;
+            return nsent;
+        }
+
+        return distributeToChildren(maxBytes, writer, pState);
+    }
+
+    private int distributeToChildren(int maxBytes, Writer writer, State pState) throws Http2Exception {
+        long oldTotalQueuedWeights = pState.totalQueuedWeights;
+        State iState = pState.poll();
+        State iNextState = pState.peek();
+        try {
+            assert iNextState == null || iNextState.pseudoTime >= iState.pseudoTime : "iNextState.pseudoTime(" +
+                    iNextState.pseudoTime + ") < " + " iState.pseudoTime(" + iState.pseudoTime + ")";
+            int nsent = distribute(iNextState == null ? maxBytes :
+                                    min(maxBytes, (int) ((iNextState.pseudoTime - iState.pseudoTime) *
+                                             ((double) iNextState.stream.weight() / oldTotalQueuedWeights)
```

Comment:
Getting rid of the parens, means you don't need to use `double`, since there is no real risk of overflow from the multiplication and the result of the division has to be rounded anyway.